### DISCUSSION TITLE
ELNES fitting of carbon-K edge

### DISCUSSION THREAD

**Discussion by PhysXu** [hyperspy/exspy#58] [Q&A]
Hi, all,

I am trying to fit one EEL spectrum by using exspy, like below:
![image](https://github.com/hyperspy/exspy/assets/35793078/923f052b-8cce-4930-95d3-55b34696d5a5)

I hope to use gaussian functions to fit the first two peaks (one is around 286 eV and the second is around 292 eV) with the fine structure as well.

**The aim is to get the center and width of the two peaks here.**

I followed the method in [Fine Structure Analysis](https://exspy.readthedocs.io/en/latest/user_guide/eels.html#id1) part of eXSpy tutorial, but the result (below) is not so good.
![image](https://github.com/hyperspy/exspy/assets/35793078/3c450d0f-6b4b-488b-a771-248e40a7396c)

Then I tried another method:
Firstly I subtract the background by using the **s.remove_background()** function;
Secondly I fit the first two peaks by using two gaussian functions separately.
Then I fix the parameters of the two gaussian peaks and fit the fine structure.
This is the results below:
![image](https://github.com/hyperspy/exspy/assets/35793078/df6ee0aa-8a7c-4cd0-bcb7-3177853731fa)

Here, I confronted a problem: I do not know how to fit the plateau between these two gaussian peaks.
And once I set all parameters free, the results went very wrong.
![image](https://github.com/hyperspy/exspy/assets/35793078/fa754fbe-10b5-4305-921b-0e8d1391e592)

I suppose that I need to add a EELSArctan function for the full fitting, but do not find such an instruction in eXSpy online tutorial.

I uploaded my jupyter file and raw data in the ZIP below.
If anyone has experience about this issue or is interested in it, any discussion is welcomed. And any suggestions will be appreciated very much!

[ELNES fit.zip](https://github.com/user-attachments/files/15848708/ELNES.fit.zip)

**Comment by PhysXu** [comment]
Hi, @ericpre , thanks very much for processing this report and moving it into the discussion channel.
This morning, I tried more on my fitting again.

I found a way to add the **Arctan** component into my fitting with the codes below:
```
arctan = epy.components.EELSArctan(x0=g1.centre.value)
m.append(arctan)
```

Now I can fit the ELNES with two fixed gaussian and one arctan, like below
![image](https://github.com/hyperspy/exspy/assets/35793078/ef6445c1-85f3-4f55-abc5-b5acd9b8a674)

However, when I tried to fit the full energy range with `m.components.C_K.fine_structure_active = True`
Stange results appeared, like below
![image](https://github.com/hyperspy/exspy/assets/35793078/ce5ca6e8-516e-4157-a9ff-72259471c9c3)

I updated my codes in the jupyter-lab code in the below zip file, along with the raw EELS data.
I am hoping to get some suggestions to improve this fitting.

[ELNES fit updated.zip](https://github.com/user-attachments/files/15856027/ELNES.fit.updated.zip)

**Comment by PhysXu** [comment]
Something updated!

I found the codes from @thomasaarholt for the fitting of Zr_L edge fine structure:
https://gist.github.com/thomasaarholt/1098819b9c930a48b3efb48df76a5185
Thanks for these codes.

I modified it for my C_K edge fitting, and get a better result.
```
def fit_C_K(s, ll=None):
    
    import exspy as epy 
    s.set_microscope_parameters(beam_energy=60,
                            convergence_angle=32,
                            collection_angle=75)
    
    m = s.create_model(low_loss = ll, GOS='gosh', auto_add_edges=False)
    m.enable_fine_structure()
    m.fit_component(m["PowerLaw"], bounded=True, signal_range=[273.5,283.5], fit_independent=True, only_current=True)
    m.assign_current_values_to_all()
    m.fit_component(m["PowerLaw"], bounded=True, signal_range=[273.5,283.5], fit_independent=True, only_current=False)


    Arctan = epy.components.EELSArctan()
    C_K_fine = epy.components.EELSCLEdge("C_K", GOS='gosh')

    C_l1 = hs.model.components1D.GaussianHF()
    C_l1.name = "C_K pai*"
    C_l1.centre.bmin = 285.
    C_l1.centre.bmax = 287.
    C_l1.height.bmin = 0
    C_l1.fwhm.bmax=1.5
    m.append(C_l1)

    m.fit_component(C_l1, optimizer='lm', signal_range=[C_l1.centre.bmin-1.0, C_l1.centre.bmax+1.0], bounded=True,)
    m.assign_current_values_to_all([C_l1])
    m.fit_component(C_l1, optimizer='lm', signal_range=[C_l1.centre.bmin-1.0, C_l1.centre.bmax+1.0], bounded=True, only_current=False)

    m.append(Arctan)
    Arctan.x0.twin = C_l1.centre
    m.fit_component(Arctan, bounded=True, signal_range=[286., 292.])
    m.assign_current_values_to_all([Arctan])
    m.fit_component(Arctan, bounded=True, signal_range=[286., 292.], only_current=False)

    C_l2 = hs.model.components1D.GaussianHF()
    C_l2.name = "C_K sigma*"
    C_l2.centre.bmin = 292.
    C_l2.centre.bmax = 293.
    C_l2.height.bmin = 0
    C_l2.fwhm.bmax=1.5

    m.append(C_l2)
    m.fit_component(C_l2, signal_range=[C_l2.centre.bmin-1.0, C_l2.centre.bmax+1.0], bounded=True,)

    m.assign_current_values_to_all([C_l2])
    m.fit_component(C_l2, signal_range=[C_l2.centre.bmin-1.0, C_l2.centre.bmax+1.0], bounded=True, only_current=False)

    m.append(C_K_fine)
    #C_K_fine.onset_energy.value = 293.  #C_l2.centre
    C_K_fine.onset_energy.twin = C_l2.centre
    #m.components.C_K.fine_structure_spline_onset = 293
    m.fit_component(C_K_fine, bounded=True, signal_range=[293., 315.])
    m.assign_current_values_to_all([C_l2])
    m.fit_component(C_K_fine, bounded=True, signal_range=[293., 315.], only_current=False)
    m.fit_component(C_l2, signal_range=[C_l2.centre.bmin-1.0, C_l2.centre.bmax+1.0], bounded=True, only_current=False)

    m.components.C_K.fine_structure_active = True
    m.fit_component(Arctan, bounded=True, signal_range=[281., 291.], only_current=False)
    m.fit_component(C_l1, signal_range=[C_l1.centre.bmin-1.0, C_l1.centre.bmax+1.0], bounded=True, only_current=False)
    m.fit_component(C_K_fine, bounded=True, signal_range=[293., 315.], only_current=False)
    m.fit_component(C_l2, signal_range=[C_l2.centre.bmin-1.0, C_l2.centre.bmax+1.0], bounded=True, only_current=False)
    print("Finished fitting Carbon_K ELNES")
    
    m.plot(plot_components=True,plot_residual=True)
    plt.xlim(280.,312.)
    
    print(m.components)
    param_list = [C_l1.centre.value, C_l1.centre.std, C_l1.fwhm.value, C_l1.fwhm.std, 
                  C_l2.centre.value, C_l2.centre.std, C_l2.fwhm.value, C_l2.fwhm.std]
    
    return m,param_list
```
![image](https://github.com/hyperspy/exspy/assets/35793078/7221936c-ca00-40b6-bc29-95ebe61e0f0b)

**The fitting error looks like that the onset_energy of C_K fine structure is not so suitable.**
I set onset_energy of C_K fine structure to be 293 eV constrainedly above.

I also tried the **twin** function of C_K onset_energy with the center of the second gaussian peak, but it went worse.
![image](https://github.com/hyperspy/exspy/assets/35793078/51088b02-a842-4bf3-9989-7e59aaf01837)

Hi, @thomasaarholt , if you see this discussion, any suggestion from you will be appreciated very much!

**Comment by ericpre** [comment]
The two approaches are similar: one is using a arctan to model the continuum, the other one is using a core edge component. 

From a quick glance, there are a few things you could do to improve the fit:
- Calibrate the energy axis: it seems that your edge start at 285 instead of 284, this will improve when using core edge component
- Use low-loss to take into account plural scattering: some of the feature from 290 eV onwards may be related to the plural scattering and this help to model this energy range instead of fitting it with the splines
- Use a narrower fitting range if you are only interested in the position of the two peaks, but this may depends on what conclusion you want to draw from that fit.
- Add more components to describe the fine structure after the second peak. If you have a good understanding of the physical meaning the feature in your edge, you could add more Gaussians, but this may be more difficult.

**Reply by PhysXu** [reply]
Hi, @ericpre , thanks very much for your suggestions.

Here is a quick reply:
1- The data has been calibrated by the zero-loss peak. Nevertheless, I will try the energy-loss position you suggested.
2- The data is from the monolayer graphene, so I suppose the plural scattering is negligible here?
3- I will try.
4- Yes, I have tried, but the fitting is hard to converge.
